hilti/toolchain/src/compiler/jit.cc

```diff
@@ -64,99 +80,288 @@ bool CxxCode::save(std::ostream& out) const {
+
+    try {
+        HILTI_DEBUG(logging::debug::Jit, util::fmt("copying library to %s", ext_lib));
+        hilti::rt::filesystem::create_directory(ext_lib.parent_path());
+        hilti::rt::filesystem::rename(_tmpdir / lib, ext_lib);
+    } catch ( hilti::rt::filesystem::filesystem_error& e ) {
+        return result::Error(e.what());
+    }
+
+    // Instantiate the library object from the file on disk, and set it up
+    // to delete the file & its directory on destruction.
+    auto library = std::shared_ptr<const Library>(new Library(ext_lib), [ext_lib](const Library* library) {
+        auto remove = library->remove();
+        if ( ! remove )
+            logger().warning(util::fmt("could not remove JIT library: %s", remove.error()));
+
+        std::error_code ec;
+        if ( ! hilti::rt::filesystem::remove(ext_lib.parent_path(), ec) )
+            logger().warning(util::fmt("could not remove JIT temporary library directory: %s", ec.message()));
+
+        delete library;
+    });
+
+    if ( _dump_code ) {
+        // Logging to driver because that's where all the other "saving to ..." messages go.
+        auto dbg = "dbg.__library__.hlto";
+        HILTI_DEBUG(logging::debug::Driver, util::fmt("saving library to %s", dbg));
+        library->save(dbg); // will go into current directory
+    }
+
+    return library;
 }
 
-std::string JIT::compilerVersion() { return detail::ClangJIT::compilerVersion(); }
+Result<JIT::JobID> JIT::_spawnJob(hilti::rt::filesystem::path cmd, std::vector<std::string> args) {
+    std::vector<std::string> cmdline = {cmd.native()};
+
+    for ( auto&& a : args )
+        cmdline.push_back(std::move(a));
 
-#else
+    auto jid = ++_job_counter;
+    HILTI_DEBUG(logging::debug::Jit, util::fmt("[job %u] %s", jid, util::join(cmdline, " ")));
 
-class detail::ClangJIT {};
+    Job& job = _jobs[jid];
+    job.process = std::make_unique<TinyProcessLib::Process>(
+        cmdline, _tmpdir, [&job](const char* bytes, size_t n) { job.stdout_ += std::string(bytes, n); },
+        [&job](const char* bytes, size_t n) { job.stderr_ += std::string(bytes, n); });
 
-JIT::JIT(std::shared_ptr<Context> context) : _context(std::move(context)) {}
+    HILTI_DEBUG(logging::debug::Jit, util::fmt("[job %u] -> pid %u", jid, job.process->get_id()));
 
-JIT::~JIT() {}
+    // TinyProcessLibProcess fails silently if there's a problem, but leaves the PID unset.
+    if ( job.process->get_id() <= 0 )
+        return result::Error(util::fmt("process failed to start: %s %s", cmd.native(), util::join(args)));
 
-bool JIT::compile() {
-    logger().error("jit: support for just-in-time compilation not available");
-    return false;
+    return jid;
 }
 
-Result<Nothing> JIT::jit() {
-    logger().error("jit: support for just-in-time compilation not available");
+Result<Nothing> JIT::_waitForJob(JobID id) {
+    if ( _jobs.find(id) == _jobs.end() )
+        return result::Error(util::fmt("unknown JIT job %u", id));
```

Comment:
This is private functionality meant to be used only to wait for previously spawned (and not yet waited for) processes. So I consider it an API misuse to use it otherwise, hence the error.  :-)